File: app/src/main/java/au/com/codeka/weather/model/CurrentCondition.java
package au.com.codeka.weather.model;

import androidx.annotation.Nullable;

import java.util.Date;

/**
 * POJO that represents the "current" conditions.
 */
public class CurrentCondition {
  private String observationLocation;
  private Date observationTime;
  @Nullable private Double temperature;
  @Nullable private Double feelsLike;
  private double precipitationLastHour;
  private double precipitationToday;
  @Nullable private Double relativeHumidity;
  private String description;
  private WeatherIcon icon;

  public String getObservationLocation() {
    return observationLocation;
  }

  public Date getObservationTime() {
    return observationTime;
  }

  @Nullable
  public Double getTemperature() {
    return temperature;
  }

  @Nullable
  public Double getFeelsLike() {
    return feelsLike;
  }

  public double getPrecipitationLastHour() {
    return precipitationLastHour;
  }

  public double getPrecipitationToday() {
    return precipitationToday;
  }

  @Nullable
  public Double getRelativeHumidity() {
    return relativeHumidity;
  }

  public String getDescription() {
    return description;
  }

  public WeatherIcon getIcon() {
    return icon;
  }

  public static class Builder {
    private CurrentCondition currentCondition;

    public Builder() {
      currentCondition = new CurrentCondition();
    }

    public Builder setObservationLocation(String location) {
      currentCondition.observationLocation = location;
      return this;
    }

    public Builder setObservationTime(Date dt) {
      currentCondition.observationTime = dt;
      return this;
    }

    public Builder setTemperature(@Nullable Double temp) {
      currentCondition.temperature = temp;
      return this;
    }

    public Builder setFeelsLike(@Nullable Double temp) {
      currentCondition.feelsLike = temp;
      return this;
    }

    public Builder setPrecipitation(@Nullable Double lastHour, @Nullable Double today) {
      currentCondition.precipitationLastHour = (lastHour == null ? 0 : lastHour);
      currentCondition.precipitationToday = (today == null ? 0 : today);
      return this;
    }

    public Builder setRelativeHumidity(@Nullable Double percent) {
      currentCondition.relativeHumidity = percent;
      return this;
    }

    public Builder setDescription(String description) {
      currentCondition.description = description;
      return this;
    }

    public Builder setIcon(WeatherIcon icon) {
      currentCondition.icon = icon;
      return this;
    }

    public CurrentCondition build() {
      return currentCondition;
    }
  }
}

File: app/src/main/java/au/com/codeka/weather/model/CurrentCondition.kt
package au.com.codeka.weather.model

import java.util.*

/** POJO that represents the "current" conditions. */
class CurrentCondition {
  var observationLocation: String? = null
    private set
  var observationTime: Date? = null
    private set
  var temperature: Double? = null
    private set
  var feelsLike: Double? = null
    private set
  var precipitationLastHour = 0.0
    private set
  var precipitationToday = 0.0
    private set
  var relativeHumidity: Double? = null
    private set
  var description: String? = null
    private set
  var icon: WeatherIcon? = null
    private set

  class Builder {
    private val currentCondition: CurrentCondition = CurrentCondition()

    fun setObservationLocation(location: String?): Builder {
      currentCondition.observationLocation = location
      return this
    }

    fun setObservationTime(dt: Date?): Builder {
      currentCondition.observationTime = dt
      return this
    }

    fun setTemperature(temp: Double?): Builder {
      currentCondition.temperature = temp
      return this
    }

    fun setFeelsLike(temp: Double?): Builder {
      currentCondition.feelsLike = temp
      return this
    }

    fun setPrecipitation(lastHour: Double?, today: Double?): Builder {
      currentCondition.precipitationLastHour = lastHour ?: 0.0
      currentCondition.precipitationToday = today ?: 0.0
      return this
    }

    fun setRelativeHumidity(percent: Double?): Builder {
      currentCondition.relativeHumidity = percent
      return this
    }

    fun setDescription(description: String?): Builder {
      currentCondition.description = description
      return this
    }

    fun setIcon(icon: WeatherIcon?): Builder {
      currentCondition.icon = icon
      return this
    }

    fun build(): CurrentCondition {
      return currentCondition
    }
  }
}